src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -118,15 +144,37 @@ internal bool OnResponseStarting()
 
         internal void FinalizeCaching()
         {
-            if (CacheResponse)
+            // Don't cache errors? 404 etc
+            if (CacheResponse && HttpContext.Response.StatusCode == 200)
             {
                 // Store the buffer to cache
-                var cacheEntry = new ResponseCachingEntry();
-                cacheEntry.StatusCode = HttpContext.Response.StatusCode;
-                foreach (var pair in HttpContext.Response.Headers)
+                var cacheEntry = new ResponseCachingEntry()
+                {
+                    Created = DateTimeOffset.UtcNow,
+                    StatusCode = HttpContext.Response.StatusCode
+                };
+
+                var headers = HttpContext.Response.Headers;
+                var count = headers.Count
+                    - (headers.ContainsKey("Date") ? 1 : 0)
+                    - (headers.ContainsKey("Content-Length") ? 1 : 0)
+                    - (headers.ContainsKey("Age") ? 1 : 0);
+                var cachedHeaders = new List<KeyValuePair<string, StringValues>>(count);
```

Comment:
Hmm the values stored in here is just thrown away later on. Also, you can just set the values directly on `cacheEntry.Headers` now.